src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -1332,31 +1332,31 @@ public static unsafe int IndexOf<T>(T[] array, T value, int startIndex, int coun
                 if (sizeof(T) == sizeof(byte))
                 {
                     int result = SpanHelpers.IndexOfValueType(
-                        ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<byte[]>(array)), startIndex),
+                        ref Unsafe.Add(ref Unsafe.As<T, byte>(ref MemoryMarshal.GetArrayDataReference(array)), startIndex),
```

Comment:
This creates an extra generic method instantiation. It should be ok - we are unlikely to hit this path for too many different Ts.

Reply:
I agree its typically fine when we can restrict the callers, but this is a general public API accessible to `Span<T>` and that can be hit for any user-defined `T`.

Given we've notably hit several aliasing bugs in the past for the JIT and had to fix them, including ones that caused invalid codegen, this seems like something we should generally be protecting against.

Reply:
3rd party code also notably looks to code in the BCL to see what it can and cannot do a lot of the time.

There are plenty of users that will see this code and copy it as is, not understanding the limitations or that they may need to fix it if the JIT ever changes.

Reply:
This code does address computation only, without dereferencing anything. The JIT should not run into any aliasing issues with it given the optimizations that it does today. In case the JIT starts doing more aggressive optimizations that depend on no aliasing, there are likely more issues to fix.

We have said many times that CoreLib is special and that it is not ok to copy code from it in general. If people copy code without understanding what it does and whether it is ok to copy it, it is their problem. I do not know how to help them.

I do not have a problem with taking this change. I am just saying that it does not make sense to fix all instances of this issue in CoreLib.

Reply:
> This code does address computation only, without dereferencing anything.

I don't believe it has to. We had general bugs in `GetArrayDataReference` which is why the generic version was made intrinsic. The non-generic version has, afaik, the same general issue since it is not intrinsic yet today.

For example we hit https://github.com/dotnet/runtime/pull/79760#issuecomment-1355285410 which was a fairly complex issue related to assertion prop and inlining caused by the aliasing of `T[]` to `RawArrayData`. There were at least 3-4 other issues that were also caused by the general aliasing.

In this case, `GetArrayDataReference` will still itself have a deref, because it's doing `ref Unsafe.As<RawData>(array).Data` and so a similar issue to the above can still occur.

> We have said many times that CoreLib is special and that it is not ok to copy code from it in general. If people copy code without understanding what it does and whether it is ok to copy it, it is their problem. I do not know how to help them.

I think the only "fix" here is to ensure that such code is annotated with an explanation of why it is ok for corelib and not elsewhere. Even within the BCL team such knowledge is not always common place for these types of specialized scenarios and may end up in unintended places due to copy/paste, refactoring, etc.

Reply:
I think fixing the non-generic `GetArrayDataReference(Array)` to be intrinsic and may be another valid fix.

Doing that would also notably also remove the `Unsafe.As` generic instantiation.

Reply:
> In this case, GetArrayDataReference will still itself have a deref, because it's doing ref Unsafe.As<RawData>(array).Data and so a similar issue to the above can still occur.

Not true anymore since it was made an intrinsic.

>I think fixing the non-generic GetArrayDataReference(Array) to be intrinsic and may be another valid fix.
>
> Doing that would also notably remove one of the Unsafe.As generic instantiations.

This would work too, I didn't do it in my GADR PR though since I wasn't sure what nodes to emit there to handle both SZ and MD arrays.

Reply:
> > This code does address computation only, without dereferencing anything. The JIT should not run into any aliasing issues with it given the optimizations that it does today

> I don't believe it has to. We had general bugs in GetArrayDataReference which is why the generic version was made intrinsic.

I meant memory writes. I believe that the aliasing bugs only occur with interleaving memory reads and writes given what the JIT can do today.

Non-generic GetArrayDataReference(Array) has to do more work to deal with MD arrays. We would have to do expansion of the non-generic GetArrayReference based on the actual type of the incoming argument to the extra work for SD arrays.

> Could you remind me what was needed to make JIT intrinsics not cause instantiations when handled by the JIT?

https://github.com/dotnet/runtime/pull/82917#discussion_r1124864890